### PR TITLE
feat: add distinct count aggregator and column distinct count

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

feat: add distinct count aggregator and column distinct count

Closes https://github.com/datafuselabs/databend/issues/8465

### CODE REVIEW

File: src/query/functions/src/aggregates/aggregate_distinct_count.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+use crate::aggregates::aggregator_common::assert_variadic_arguments;
+
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct AggregateCountState {
+    hll: HyperLogLog<DataValue>,
+    input_rows: usize,
+}
+
+#[derive(Clone)]
+pub struct AggregateDistinctCountFunction {
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_distinct_count.rs:37]: We already have `AggregateDistinctCountFunction` as combinator, should rename to `ApproxCountDistinctFunction`

File: src/query/functions/src/aggregates/aggregator.rs [hunk 2]
```diff
@@ -56,6 +57,7 @@ impl Aggregators {
         factory.register("uniq", aggregate_combinator_uniq_desc());
 
         factory.register("retention", aggregate_retention_function_desc());
+        factory.register("distinct_counts", AggregateDistinctCountFunction::desc());
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregator.rs:60]: approx_count_distinct

File: src/query/storages/fuse/fuse/src/fuse_table.rs [hunk 3]
```diff
@@ -476,20 +476,19 @@ impl Table for FuseTable {
 }
 
 #[derive(Default)]
-struct FakedColumnStatisticsProvider {
+struct FuseColumnStatisticsProvider {
     column_stats: HashMap<ColumnId, FuseColumnStatistics>,
-    // faked value, just the row number
-    faked_ndv: u64,
+    row_count: u64,
```
Comment (sundy-li) [inline src/query/storages/fuse/fuse/src/fuse_table.rs:481]: It's `distinct_count`

Comment (sundy-li) [reply src/query/storages/fuse/fuse/src/fuse_table.rs:481]: or `ndv`

File: src/query/storages/fuse/fuse/src/fuse_table.rs [hunk 4]
```diff
@@ -476,20 +476,19 @@ impl Table for FuseTable {
 }
 
 #[derive(Default)]
-struct FakedColumnStatisticsProvider {
+struct FuseColumnStatisticsProvider {
```
Comment (sundy-li) [inline src/query/storages/fuse/fuse/src/fuse_table.rs:479]: we should keep the name cause it's faked.

File: src/query/functions/src/aggregates/aggregate_distinct_count.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+    fn state_layout(&self) -> Layout {
+        Layout::new::<AggregateCountState>()
+    }
+
+    fn accumulate(
+        &self,
+        place: StateAddr,
+        columns: &[ColumnRef],
+        _validity: Option<&Bitmap>,
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_distinct_count.rs:88]: _validity should be considered, we can't take null into count.

File: src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct AggregateApproximateDistinctCountState {
+    hll: HyperLogLog<DataValue>,
+    input_rows: u64,
+}
+
+#[derive(Clone)]
+pub struct AggregateApproximateDistinctCountFunction {
+    display_name: String,
+    nullable: bool,
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs:39]: useless field.

File: src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    fn get_own_null_adaptor(
+        &self,
+        _nested_function: super::AggregateFunctionRef,
+        _params: Vec<DataValue>,
+        _arguments: Vec<DataField>,
+    ) -> Result<Option<super::AggregateFunctionRef>> {
+        let mut f = self.clone();
+        f.nullable = true;
+        Ok(Some(Arc::new(f)))
+    }
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs:168]: should be removed.

File: src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+use super::aggregate_function::AggregateFunction;
+use super::aggregate_function_factory::AggregateFunctionDescription;
+use super::StateAddr;
+use crate::aggregates::aggregator_common::assert_variadic_arguments;
+
+#[derive(serde::Serialize, serde::Deserialize)]
+pub struct AggregateApproximateDistinctCountState {
+    hll: HyperLogLog<DataValue>,
+    input_rows: u64,
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs:33]: What's the usage of input_rows?

File: src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        let builder: &mut MutablePrimitiveColumn<u64> = Series::check_get_mutable_column(array)?;
+        let state = place.get::<AggregateApproximateDistinctCountState>();
+        let input_rows = state.input_rows as f64;
+        let len = state.hll.len();
+        if state.input_rows == 0 {
+            builder.append_value(0);
+        } else if input_rows > len {
+            builder.append_value(len as u64);
+        } else {
+            builder.append_value((len / input_rows) as u64);
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs:154]: why ?

File: src/query/functions-v2/src/aggregates/aggregate_count.rs [hunk 10]
```diff
@@ -32,7 +32,7 @@ use super::aggregate_function_factory::AggregateFunctionDescription;
 use super::StateAddr;
 use crate::aggregates::aggregator_common::assert_variadic_arguments;
 
-pub struct AggregateCountState {
+pub struct AggregateApproximateDistinctCountState {
```
Comment (sundy-li) [inline src/query/functions-v2/src/aggregates/aggregate_count.rs:35]: It's in function-v2, we can't modify this.

File: src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                Some(b) => b.unset_bits(),
+                None => 0,
+            };
+            if column.len() == nulls {
+                continue;
+            }
+            column.to_values().iter().for_each(|value| {
+                state.hll.push(value);
+            });
+        }
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs:102]: It's safe to use `columns[0]`

File: src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    display_name: String,
+}
+
+impl AggregateApproximateDistinctCountFunction {
+    pub fn try_create(
+        display_name: &str,
+        _params: Vec<DataValue>,
+        arguments: Vec<DataField>,
+    ) -> Result<Arc<dyn AggregateFunction>> {
+        assert_variadic_arguments(display_name, arguments.len(), (0, 1))?;
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs:47]: must be one argument.

File: src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs (hunk 11 above)
Comment (sundy-li) [reply src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs:102]: we must check if the `validity` has NULL values.

File: src/query/storages/fuse/fuse/src/statistics/column_statistic.rs [hunk 13]
```diff
@@ -23,6 +26,27 @@ use common_storages_index::SupportedType;
...
+    column: &Arc<dyn Column>,
+    column_field: ColumnWithField,
+) -> Result<u64> {
+    let distinct_values = eval_aggr(
+        "approx_count_distinct",
+        vec![],
+        &[column_field],
+        column.len(),
+    )?;
+    if distinct_values.len() > 0 {
```
Comment (sundy-li) [inline src/query/storages/fuse/fuse/src/statistics/column_statistic.rs:39]: no need to check the len, `distinct_values.len() > 0` is ensured even in empty set.

```
MySQL [(none)]> select max(number) from numbers(3) where number > 333333;
+-------------+
| max(number) |
+-------------+
|        NULL |
+-------------+
1 row in set (0.009 sec)
```

File: src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs [hunk 14]
```diff
@@ -44,7 +45,7 @@ impl AggregateApproximateDistinctCountFunction {
         _params: Vec<DataValue>,
         arguments: Vec<DataField>,
     ) -> Result<Arc<dyn AggregateFunction>> {
-        assert_variadic_arguments(display_name, arguments.len(), (0, 1))?;
+        assert_variadic_arguments(display_name, arguments.len(), (1, 32))?;
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_approximate_distinct_count.rs:48]: assert_unary_arguments

**Review by sundy-li — APPROVED** [review]

**Comment by BohuTANG** [comment]
@soyeric128 Pls add the new func: `approx_count_distinct` function, thank you!

**Pull request merged**